synapse/storage/roommember.py

```diff
@@ -639,6 +661,39 @@ def f(txn):
         count = yield self.runInteraction("did_forget_membership", f)
         defer.returnValue(count == 0)
 
+    @cached()
+    def get_forgotten_rooms_for_user(self, user_id):
+        """Gets all rooms the user has forgotten.
+
+        Args:
+            user_id (str)
+
+        Returns:
+            Deferred[set[str]]
+        """
+
+        def _get_forgotten_rooms_for_user_txn(txn):
+            # This is a slightly convoluted query that first looks up all rooms
+            # that the user has forgotten in the past, then rechecks that list
+            # to see if any have subsequently been updated. This is done so that
+            # we can use a partial index on `forgotten = 1` on the assumption
+            # that few users will actually forget many rooms.
+            sql = """
+                SELECT room_id, (
+                    SELECT count(*) FROM room_memberships
+                    WHERE room_id = m.room_id AND user_id = m.user_id AND forgotten = 0
+                ) AS count
+                FROM room_memberships AS m
+                WHERE user_id = ? AND forgotten = 1
```

Comment:
this is going to be very painful until that background reindex completes, isn't it.

Reply:
It takes about ~500ms to 900ms for large accounts, small accounts take a few ms.

The query plan does a bitmap scan across `room_id` and `user_id` indices:

```
matrix=> explain analyze SELECT room_id, (
                    SELECT count(*) FROM room_memberships
                    WHERE room_id = m.room_id AND user_id = m.user_id AND forgotten = 0
                ) AS count
                FROM room_memberships AS m
                WHERE user_id = '@e:matrix.org' AND forgotten = 1
                GROUP BY room_id, user_id;
                                                                         QUERY PLAN                                                                          
-------------------------------------------------------------------------------------------------------------------------------------------------------------
 Group  (cost=2518.90..4262.71 rows=4 width=68) (actual time=0.412..0.412 rows=0 loops=1)
   Group Key: m.room_id, m.user_id
   ->  Sort  (cost=2518.90..2518.91 rows=4 width=60) (actual time=0.411..0.411 rows=0 loops=1)
         Sort Key: m.room_id
         Sort Method: quicksort  Memory: 25kB
         ->  Index Scan using room_memberships_user_id on room_memberships m  (cost=0.57..2518.86 rows=4 width=60) (actual time=0.407..0.407 rows=0 loops=1)
               Index Cond: (user_id = '@e:matrix.org'::text)
               Filter: (forgotten = 1)
               Rows Removed by Filter: 46
   SubPlan 1
     ->  Aggregate  (cost=435.94..435.95 rows=1 width=8) (never executed)
           ->  Bitmap Heap Scan on room_memberships  (cost=433.92..435.93 rows=1 width=0) (never executed)
                 Recheck Cond: ((user_id = m.user_id) AND (room_id = m.room_id))
                 Filter: (forgotten = 0)
                 ->  BitmapAnd  (cost=433.92..433.92 rows=1 width=0) (never executed)
                       ->  Bitmap Index Scan on room_memberships_user_id  (cost=0.00..32.24 rows=1289 width=0) (never executed)
                             Index Cond: (user_id = m.user_id)
                       ->  Bitmap Index Scan on room_memberships_room_id  (cost=0.00..401.43 rows=17981 width=0) (never executed)
                             Index Cond: (room_id = m.room_id)
 Planning time: 0.144 ms
 Execution time: 0.468 ms
(21 rows)
```